src/test/java/com/bravo/user/controller/PaymentControllerTest.java

```diff
@@ -0,0 +1,82 @@
+package com.bravo.user.controller;
+
+import com.bravo.user.App;
+import com.bravo.user.model.dto.PaymentDto;
+import com.bravo.user.service.PaymentService;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.ResultActions;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@ContextConfiguration(classes = { App.class })
+@ExtendWith(SpringExtension.class)
+@SpringBootTest()
+@AutoConfigureMockMvc
+public class PaymentControllerTest {
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private PaymentService paymentService;
+
+    private List<PaymentDto> payments;
+
+    @BeforeEach
+    public void beforeEach() {
+        final List<Integer> ids = IntStream.range(1, 10).boxed().collect(Collectors.toList());
+
+        this.payments = ids.stream().map(id -> createPaymentDto(Integer.toString(id)))
+                .collect(Collectors.toList());
+    }
+
+    @Test
+    void getRetrieveByUserId() throws Exception {
+        final String userId = "123a-456b";
+
+        when(paymentService.retrieveByUserId(anyString())).thenReturn(payments);
+
+        final ResultActions result = this.mockMvc.perform(get("/payment/retrieve/".concat(userId)))
+                .andExpect(status().isOk());
+
+        for (int i = 0; i < payments.size(); i++) {
+            result.andExpect(jsonPath(String.format("$[%d].id", i)).value(payments.get(i).getId()));
+        }
+
+        verify(paymentService).retrieveByUserId(userId);
+    }
+
+    @Test
+    void getRetrieveByUserId_Space() throws Exception {
+        this.mockMvc.perform(get("/payment/retrieve/ /")).andExpect(status().isBadRequest());
+    }
+
+    @Test
+    void getRetrieveByUserId_Missing() throws Exception {
+        this.mockMvc.perform(get("/address/retrieve")).andExpect(status().isNotFound());
```

Comment:
Great catch. I definitely missed this and should've thought it out more. I think this would be a proper test case for testing for a call of get all payments regardless of userid. That is not a request mapping that was created so it does return a 404. The proper test for this (an id that doesn't exists) would be something like "/payment/retrieve/foo" and would expect this the be a valid request and just returns empty list.

I'm also noticing I used "/address/retrieve" here and not "payment/retrieve". My mistake there, I should not have let that through.

On the topic of valid user ids, something I also considered when going through this exercise was creating a payment validator similar to user validator. I opted for using the userid validator that already existed instead because in this exercise we aren't creating payments.